src/index.ts

```diff
@@ -854,6 +869,98 @@ export abstract class PinejsClientCore<PinejsClient> {
 		}
 	}
 
+	private waitAsync(delayMs: number): Promise<void> {
+		return new Promise<void>((resolve) => {
+			setTimeout(resolve, delayMs);
+		});
+	}
+
+	private canRetryDefaultHandler(err: any) {
+		const code = err?.statusCode ?? 0;
+		return code === 0 || code === 429 || (code >= 500 && code < 600);
+	}
+
+	protected async callWithDelay<T>(
+		fnCall: () => Promise<T>,
+		delayMs: number,
+		attempt: number,
+		isLastAttempt: boolean,
+		prevErr?: any,
+		fnLogger?: (
+			prevErr: any,
+			delayMs: number,
+			attempt: number,
+			isLastAttempt: boolean,
+		) => void,
+	): Promise<T> {
+		if (!delayMs) {
+			return await fnCall();
+		}
+
+		fnLogger?.(prevErr, delayMs, attempt, isLastAttempt);
+		await this.waitAsync(delayMs);
+		return fnCall();
+	}
+
+	protected async callWithRetry<T>(
+		fnCall: () => Promise<T>,
+		retryParameters: RetryParameters,
+	): Promise<T> {
+		const minDelayMs =
+			retryParameters.minDelayMs ?? this.retryParameters.minDelayMs;
+		const maxDelayMs =
+			retryParameters.maxDelayMs ??
+			this.retryParameters.maxDelayMs ??
+			minDelayMs;
+		const maxAttempts =
+			retryParameters.maxAttempts ?? this.retryParameters.maxAttempts ?? 0;
+		const logHandler = retryParameters.logger ?? this.retryParameters.logger;
+
+		let attempt = 0;
+		let delay = 0;
+		let prevErr: any;
+
+		if (minDelayMs && (maxDelayMs || maxAttempts)) {
+			const canRetryHandler =
+				retryParameters.canRetry ??
+				this.retryParameters.canRetry ??
+				this.canRetryDefaultHandler;
+
+			// Go up to the max number of attempts or else if
+			// maxAttempts not specified, then go up to the max delay
+			while (
+				attempt < maxAttempts ||
+				(maxAttempts <= 0 && delay < maxDelayMs!)
+			) {
+				try {
+					return await this.callWithDelay(
+						fnCall,
+						delay,
+						attempt,
+						false,
+						prevErr,
+						logHandler,
+					);
+				} catch (err) {
+					if (!canRetryHandler(err)) {
+						throw err;
+					}
+					prevErr = err;
+				}
+				delay = Math.min(2 ** attempt * minDelayMs, maxDelayMs!);
```

Comment:
For reference this will cause problems if `maxDelayMs` is actually undefined (and it can be):
```ts
Math.min(1000, undefined) // NaN
```
I think this will be solved by treating max attempts/max delay as separate things but if not then this would definitely need solving

Reply:
@Page- `maxDelayMs` should never be `undefined`
See line
```
		const maxDelayMs =
			retryParameters.maxDelayMs ??
			this.retryParameters.maxDelayMs ??
			minDelayMs;
```